compiler/rustc_passes/src/reachable.rs

```diff
@@ -205,19 +206,34 @@ impl<'tcx> ReachableContext<'tcx> {
                     }
 
                     hir::ItemKind::Const(_, _, _, init) => {
-                        // Only things actually ending up in the final constant value are reachable
-                        // for codegen. Everything else is only needed during const-eval, so even if
-                        // const-eval happens in a downstream crate, all they need is
-                        // `mir_for_ctfe`.
+                        if self.tcx.generics_of(item.owner_id).own_requires_monomorphization() {
+                            // In this case, we don't want to evaluate the const initializer.
+                            // In lieu of that, we have to consider everything mentioned in it
+                            // as reachable, since it *may* end up in the final value.
+                            self.visit_nested_body(init);
+                            return;
+                        }
+
+                        let predicates = self.tcx.predicates_of(item.owner_id);
+                        let predicates = predicates.instantiate_identity(self.tcx).predicates;
+                        if traits::impossible_predicates(self.tcx, predicates) {
+                            // The constant is impossible to reference.
```

Comment:
Extend comment: If the usage site also assumes these impossible bounds (so of its ParamEnv also contains these bounds), it's still fine because the body gets replaced with `Unreachable` by MIR pass `ImpossiblePredicates`.

We might want to add a test for all of this.

Note to self: However, there are cases where rhe MIR pass and reachability disagrees about what constitutes an impossible bound (namely if the predicates contains early-bound regions). Check if this is problematic and can be exploited.